Api creation

FayToward opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
If this PR closes an issue, add '<AB#213700>' somewhere in the PR summary. As a minimum, please *always* link to the relevant work items e.g. AB#213700 (work item number in Azure DevOps). Follow the format below carefully, guidance found here: https://learn.microsoft.com/en-us/azure/devops/boards/github/link-to-from-github?view=azure-devops. Note: The Title pattern should be `{work item number}: {title}` -->

# **What this PR does / why we need it**:
*A brief description of changes being made*
*Link to the relevant work items: e.g: Relates to ADO Work Item AB#213700 and builds on #3376 (link to ADO Build ID URL)*

# **Special notes for your reviewer**
*Any specific actions or notes on review?*

# Testing
*Any relevant testing information and pipeline runs.*

# Checklist (please delete before completing or setting auto-complete)
- [ ] Story Work items associated (not Tasks)
- [ ] Successful testing run(s) link provided
- [ ] Title pattern should be `{work item number}: {title}`
- [ ] Description covers all the changes in the PR
- [ ] This PR contains documentation
- [ ] This PR contains tests
- [ ] Version number in **package.json** has been updated

# **How does this PR make you feel**:
![gif]([https://giphy.com/)

FayToward committed:
Add preference GraphQL type and schema

FayToward committed:
Update Dockerfile and package versions

FayToward committed:
Update preference routes and docker-compose ports

FayToward committed:
reverting file

FayToward committed:
Refactor preference route handlers to use async/await

FayToward committed:
Update port number to 3009 in appConfig.yaml and helm/ffc-sfd-permissions-api/values.yaml

FayToward committed:
Add database connection and dependencies

FayToward committed:
Update PostgreSQL port in docker-compose.override.yaml

johnwatson484 reviewed (commented):
There's an issue with the permissions on the `start` and `test` scripts in the repo.  Essentially a bug with the ADP scaffolding, but we can solve this by deleting them and copying them from another working repo.  Such as this example -> https://github.com/DEFRA/ffc-pay-processing/tree/main/scripts (just remember to rename the service reference in the script!)

johnwatson484 commented on Dockerfile:
```diff
@@ -1,6 +1,6 @@
-ARG PORT=3000
-ARG PORT_DEBUG=9229
+ARG PORT=3008
+ARG PORT_DEBUG=9008
```
I think this one should stay as `9229` so the default node port is used.  I'm not sure if there's anything else that would break if we change it.  So suggest keeping consistent with others internal to the pod.

johnwatson484 commented on app/graphql/preference.js:
```diff
@@ -0,0 +1,11 @@
+const { GraphQLObjectType, GraphQLString, GraphQLInt, GraphQLList } = require('graphql')
+
+const PreferenceType = new GraphQLObjectType({
+  name: 'Preference',
+  fields: {
+    sbi: { type: GraphQLInt },
+    preference: { type: GraphQLString }
+  }
+})
+
+module.exports = { PreferenceType }
```
You can delete the `graphql` folder, it's not being used for anything.

johnwatson484 commented on app/routes/preference.js:
```diff
@@ -0,0 +1,18 @@
+    method: 'GET',
+    path: '/preference/{sbi}',
+      handler: async (request, h) => {
+        return h.response(await db.Preference.findOne({ where: { sbi: request.query.sbi } }))
```
`Preference` should be `preference` as it needs to match the case in your model definition (app/models/preferences)

![image](https://github.com/DEFRA/ffc-sfd-permissions-api/assets/42574977/c0adc129-d55b-434a-8bf2-019ff622c720)

johnwatson484 replied:
Also, my bad for the example I gave you to reference.  But think it should be `request.param.sbi` not `request.query.sbi`

`param` refers to a value in the url path.  eg `/preferences/123456789` whereas `query` refers to a value in the query string. eg `/preferences?sbi=123456789`

For info, there's also `request.payload.sbi` which would take it from the body of the request when using `POST`

eg `/preferences` with a payload `{ sbi: 123456789 }`

johnwatson484 commented on app/routes/preference.js:
```diff
@@ -0,0 +1,18 @@
+const db = require('../data')
```
There are linting errors in this file which cause the tests to fail.

Make sure that you've got `standard` installed so it gives you fast feedback on these as you write code.  Also means you can turn on "fix on save" so to fix everything you just need to click save or do `CTRL+S`

You can install it by running `npm i -g standard` and it will install it globally on your machine so every repo can take advantage of it.  `-g` is the global bit, otherwise it would add it to the `package.json` of the repo only (which it should be there anyway)

johnwatson484 commented on changelog/db.changelog-1.0.xml:
```diff
@@ -14,4 +14,4 @@
         </createTable>
     </changeSet>
-</databaseChangeLog>
+</databaseChangeLog>
```
There's still a difference in this file that I don't think it will like as you can't change it once run.  Looks like it's that you've removed an empty line at the end.

johnwatson484 commented on docker-compose.override.yaml:
```diff
@@ -19,9 +19,9 @@ services:
   ffc-sfd-permissions-api-postgres:
     ports:
-      - "5432:5432"
+      - "5441:5432"
```
Been thinking of this.  What if we set the port to be `5008`?  Then we're consistent with our port convention above?

Slightly wary that we may come across a port number at some point we can't do that for though.

johnwatson484 commented on appConfig/appConfig.yaml:
```diff
@@ -1,6 +1,6 @@
   value: "production"
 - key: PORT
-  value: "3000"
+  value: "3009"
```
This should be `3008` to match what is set in `Dockerfile`

johnwatson484 commented on helm/ffc-sfd-permissions-api/values.yaml:
```diff
@@ -14,12 +14,12 @@ deployment: {}
 container:
   memCpuTier: L
-  port: 3000
+  port: 3009
```
This should be `3008` to match the Dockerfile as that is what the container runs under

johnwatson484 commented on helm/ffc-sfd-permissions-api/values.yaml:
```diff
@@ -14,12 +14,12 @@ deployment: {}
 livenessProbe:
   path: /healthz
-  port: 3000
+  port: 3009
```
`3008`

johnwatson484 commented on helm/ffc-sfd-permissions-api/values.yaml:
```diff
@@ -28,7 +28,7 @@ livenessProbe:
 readinessProbe:
   path: /healthy
-  port: 3000
+  port: 3009
```
`3008`

johnwatson484 commented on package.json:
```diff
@@ -20,7 +20,11 @@
-    "applicationinsights": "2.8.0"
+    "applicationinsights": "2.8.0",
+    "pg": "8.7.3",
+    "pg-hstore": "2.3.4",
```
Incidentally, `pg-hstore` is required as a dependency of `sequelize` so does need to be included.

https://sequelize.org/docs/v6/getting-started/

![image](https://github.com/DEFRA/ffc-sfd-permissions-api/assets/42574977/357f9cf3-5662-4d10-8fb6-4c4def08fd8b)

FayToward committed:
Update Dockerfile and database configuration

FayToward committed:
Update preference route and port number***

***Update appConfig.yaml port number***

***Update docker-compose.override.yaml port number***

***Update helm values.yaml port number

FayToward committed:
Update start and test scripts

johnwatson484 commented on app/models/preferences.js:
```diff
@@ -0,0 +1,11 @@
+module.exports = (sequelize, DataTypes) => {
+  return sequelize.define('preferences', {
```
This should be `preference` rather than `preferences` as needs to match what follows `db.` in `routes/preferences`

We could debate over whether these should be single or plural but my view is they represent a model rather than a collection so singular makes sense.

FayToward committed:
Refactor function to improve performance

FayToward committed:
Fix formatting issue in db.changelog-1.0.xml

FayToward committed:
Update file permissions for scripts/start and scripts/test

FayToward committed:
Update file permissions for database migration scripts

FayToward committed:
Change file mode for postgres-wait script

FayToward committed:
Change column type from VARCHAR(255) to INTEGER in backend table

FayToward committed:
Update column "name" type to VARCHAR(10) in db.changelog-1.0.xml

FayToward committed:
Update column types in database changelogs

FayToward committed:
Update network name in docker-compose file

FayToward committed:
Update network name in docker-compose file

FayToward committed:
Update databaseConfig import in index.js and fix request.param.sbi typo in preference.js

FayToward committed:
Refactor preference route to return response with data object

FayToward committed:
Merge branch 'main' into api-creation
